source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -316,6 +310,98 @@ void RedisCluster::RedisDiscoverySession::startResolveRedis() {
   current_request_ = client->client_->makeRequest(ClusterSlotsRequest::instance_, *this);
 }
 
+void RedisCluster::RedisDiscoverySession::updateDnsStats(
+    Network::DnsResolver::ResolutionStatus status, bool empty_response) {
+  if (status == Network::DnsResolver::ResolutionStatus::Failure || empty_response) {
+    if (status == Network::DnsResolver::ResolutionStatus::Failure) {
+      parent_.info_->stats().update_failure_.inc();
+    } else {
+      parent_.info_->stats().update_empty_.inc();
+    }
+  }
+}
+
+void RedisCluster::RedisDiscoverySession::resolveClusterHostnames(ClusterSlotsPtr&& slots) {
+  std::shared_ptr<ClusterSlotsPtr> s = std::make_shared<ClusterSlotsPtr>(std::move(slots));
+  resolveHostname(s, 0);
+}
+
+void RedisCluster::RedisDiscoverySession::resolveHostname(std::shared_ptr<ClusterSlotsPtr> slots,
+                                                          std::size_t index) {
+  if (index >= (**slots).size()) {
+    finishClusterHostnameResolution(std::move(slots));
+    return;
+  }
+  auto& slot = (**slots)[index];
+  if (slot.primary() == nullptr) {
+    ENVOY_LOG(trace,
+              "starting async DNS resolution for primary slot address {} at index location {}",
+              slot.primary_hostname_, index);
+    parent_.dns_resolver_->resolve(
+        slot.primary_hostname_, parent_.dns_lookup_family_,
+        [this, index,
+         slots = std::move(slots)](Network::DnsResolver::ResolutionStatus status,
+                                   std::list<Network::DnsResponse>&& response) -> void {
+          auto& slot = (**slots)[index];
+          ENVOY_LOG(trace, "async DNS resolution complete for {}", slot.primary_hostname_);
+          updateDnsStats(status, response.empty());
+          // If DNS resolution for a primary fails, we stop resolution for remaining, and reset the
+          // timer.
+          if (status != Network::DnsResolver::ResolutionStatus::Success) {
+            ENVOY_LOG(error, "Unable to resolve cluster slot primary hostname {}",
+                      slot.primary_hostname_);
+            resolve_timer_->enableTimer(parent_.cluster_refresh_rate_);
+            return;
+          }
+          // Primary slot address resolved
+          slot.setPrimary(Network::Utility::getAddressWithPort(
+              *response.front().addrInfo().address_, slot.primary_port_));
+          // Continue on to resolve replicas
+          resolveReplicas(std::move(slots), index);
```

Comment:
If we want to resolve all the replicas, why don't we request a DNS resolution for all of them in parallel? I think that might avoid requiring this recursive logic and it would also potentially speed up resolution?

Reply:
Good call. I moved the logic into a loop. There's still an aspect of recursion, but that calls into `resolveHostname`.

Reply:
Why do we still need a recursive aspect? I'm still having a hard time following the logic here, so I'd love to simplify if possible